vowpalwabbit/core/include/vw/core/learner.h

```diff
@@ -482,10 +482,18 @@ class common_learner_builder
     this->learner_ptr->_cleanup_example_f = [fn_ptr, data](polymorphic_ex ex) { fn_ptr(*data, ex); };
   )
 
+  // Set the function pointer for persisting metrics for the learner.
+  // This function enforces that force reductions can only add metrics in their own namespace.
+  // The metrics output dictionary is a dictionary of dictionaries, where each sub-dictionary
+  // contains the metrics for a single learner. The key for each sub-dictionary is the learner's name.
   LEARNER_BUILDER_DEFINE(set_persist_metrics(void (*fn_ptr)(DataT&, metric_sink&)),
     assert(fn_ptr != nullptr);
     DataT* data = this->learner_data.get();
-    this->learner_ptr->_persist_metrics_f = [fn_ptr, data](metric_sink& metrics) { fn_ptr(*data, metrics); };
+    std::string learner_name = this->name;
+    this->learner_ptr->_persist_metrics_f = [fn_ptr, data, learner_name](metric_sink& metrics) {
+        metrics[learner_name] = {};
```

Comment:
here you should be able to tell `metric_sink` that it should push this new context called `learner_name`

something like
```
{
 metrics.push_context(learner_name);
 fn_ptr(*data, metrics);
 metrics.pop_context();
}
```

back inside any fn_ptr() you should be able to only operate inside this subcontext of the metrics object - you should not be able to access other contexts that might also exist.

Reply:
Thanks for the review. This makes sense.

Reply:
Hi @lalo, I am having a question. I did not find any references for `push_context` and `pop_context` methods in the `VW::metric_sink` class. Am I missing something over here? Will we have to define it?